File: cpp/tests/environmental/bme280_test_linux/bme280_test_linux.cpp
#ifndef TEST_I2C_BUS
#define TEST_I2C_BUS 1
#endif
#ifndef TEST_ADDR
#define TEST_ADDR 0x76
#endif

#include <stdio.h>
#include "I2CTransportLinux.h"
#include "BME280.h"

static int passed = 0, failed = 0;

static void check_true(bool cond, const char *label) {
    if (cond) { printf("PASS %s\n", label); passed++; }
    else       { printf("FAIL %s\n", label); failed++; }
}

int main() {
    I2CTransportLinux transport(TEST_I2C_BUS, TEST_ADDR);

    BME280Minimal bme(transport);

    bme._dig_T1 = 27504;
    bme._dig_T2 = 26435;
    bme._dig_T3 = -1000;
    bme._dig_P1 = 36477;
    bme._dig_P2 = -10685;
    bme._dig_P3 = 3024;
    bme._dig_P4 = 2855;
    bme._dig_P5 = 140;
    bme._dig_P6 = -7;
    bme._dig_P7 = 15500;
    bme._dig_P8 = -14600;
    bme._dig_P9 = 6000;

    float t = bme._compensate_temp(519888);
    check_true(t > 24.0f && t < 26.0f, "temp_compensation");

    float p = bme._compensate_pressure(415148);
    check_true(p > 1005.0f && p < 1008.0f, "pressure_compensation");

    bme._dig_H1 = 75;
    bme._dig_H2 = 362;
    bme._dig_H3 = 0;
    bme._dig_H4 = 341;
    bme._dig_H5 = 50;
    bme._dig_H6 = 30;
    float h = bme._compensate_humidity(29000);
    check_true(h > 30.0f && h < 70.0f, "humidity_compensation");

    BME280Full bme_full(transport);

    float t2 = bme_full.temperature();
    check_true(t2 >= -40.0f && t2 <= 85.0f, "temperature_range");

    float p2 = bme_full.pressure();
    check_true(p2 >= 300.0f && p2 <= 1100.0f, "pressure_range");

    float h2 = bme_full.humidity();
    check_true(h2 >= 0.0f && h2 <= 100.0f, "humidity_range");

    bme_full.set_oversampling(BME280Full::OSRS_X4, BME280Full::OSRS_X2, BME280Full::OSRS_X1);
    check_true(bme_full._osrs_t == 3 && bme_full._osrs_p == 2 && bme_full._osrs_h == 1, "set_oversampling");

    float alt = bme_full.altitude();
    check_true(alt >= -500.0f && alt <= 9000.0f, "altitude_range");

    float slp = bme_full.sea_level_pressure(0.0f);
    check_true(slp >= 900.0f && slp <= 1100.0f, "sea_level_pressure");

    check_true(bme_full.chip_id() == 0x60, "chip_id");

    bme_full.reset();
    check_true(true, "reset");

    printf("===DONE: %d passed, %d failed===\n", passed, failed);
    return failed == 0 ? 0 : 1;
}

File: python/tests/environmental/bme280_test_linux.py
import os
from periph.transport.i2c_linux import I2CTransport
from periph.chips.environmental.bme280 import BME280Minimal, BME280Full

passed = 0
failed = 0

I2C_BUS = int(os.environ.get('LINUX_I2C_BUS', '1'))
I2C_ADDR = int(os.environ.get('I2C_ADDR', '0x76'), 16)

transport = I2CTransport(I2C_BUS, I2C_ADDR)

bme = BME280Minimal(transport)

bme._dig_T1 = 27504
bme._dig_T2 = 26435
bme._dig_T3 = -1000
bme._dig_P1 = 36477
bme._dig_P2 = -10685
bme._dig_P3 = 3024
bme._dig_P4 = 2855
bme._dig_P5 = 140
bme._dig_P6 = -7
bme._dig_P7 = 15500
bme._dig_P8 = -14600
bme._dig_P9 = 6000

t = bme._compensate_temp(519888)
if abs(t - 25.08) < 0.1:
    print('PASS temperature_compensation')
    passed += 1
else:
    print('FAIL temperature_compensation: expected 25.08, got {}'.format(t))
    failed += 1

p = bme._compensate_pressure(415148)
if abs(p - 1006.53) < 0.5:
    print('PASS pressure_compensation')
    passed += 1
else:
    print('FAIL pressure_compensation: expected 1006.53, got {}'.format(p))
    failed += 1

bme._dig_H1 = 75
bme._dig_H2 = 362
bme._dig_H3 = 0
bme._dig_H4 = 341
bme._dig_H5 = 50
bme._dig_H6 = 30
bme._t_fine = 128422

h = bme._compensate_humidity(29000)
if 30.0 <= h <= 70.0:
    print('PASS humidity_compensation')
    passed += 1
else:
    print('FAIL humidity_compensation: expected 30-70 %RH, got {}'.format(h))
    failed += 1

bme_full = BME280Full(transport)
bme_full._mode = 0
if bme_full._osrs_t == 1 and bme_full._osrs_p == 1 and bme_full._osrs_h == 1:
    print('PASS default_oversampling')
    passed += 1
else:
    print('FAIL default_oversampling: got osrs_t={} osrs_p={} osrs_h={}'.format(
        bme_full._osrs_t, bme_full._osrs_p, bme_full._osrs_h))
    failed += 1

bme_full.set_oversampling(BME280Full.OSRS_X4, BME280Full.OSRS_X2, BME280Full.OSRS_X1)
if bme_full._osrs_t == 3 and bme_full._osrs_p == 2 and bme_full._osrs_h == 1:
    print('PASS set_oversampling')
    passed += 1
else:
    print('FAIL set_oversampling: got osrs_t={} osrs_p={} osrs_h={}'.format(
        bme_full._osrs_t, bme_full._osrs_p, bme_full._osrs_h))
    failed += 1

alt = bme_full.altitude(1013.25)
if alt >= -500 and alt <= 9000:
    print('PASS altitude')
    passed += 1
else:
    print('FAIL altitude: expected valid range, got {}'.format(alt))
    failed += 1

slp = bme_full.sea_level_pressure(0)
if slp >= 900 and slp <= 1100:
    print('PASS sea_level_pressure')
    passed += 1
else:
    print('FAIL sea_level_pressure: expected ~1013, got {}'.format(slp))
    failed += 1

transport.close()
print('===DONE: {} passed, {} failed==='.format(passed, failed))
